Enable bulk registration (for use with HOOK_CHAIN=yes)

pc-fan opened the pull request:
Pull request for #10 

This is intended to be used with the following option in the dehydrated confiq file:
```
HOOK_CHAIN="yes"
```

For information about HOOK_CHAIN please see https://github.com/lukas2511/dehydrated/blob/master/docs/hook_chain.md

The changes worked for me and should be backwards compatible for use without HOOK_CHAIN. Maybe you could have a look at it @jbjonesjr?

pc-fan committed:
Enable bulk registration (for use with HOOK_CHAIN=yes). See #10

jbjonesjr commented on manual_hook.rb:
```diff
@@ -2,56 +2,75 @@
+  challenges = []
+  while ARGV.length >= 3
+    domain = ARGV.shift
+    acme_domain = "_acme-challenge.#{domain}"
```
`acme_domain = "_acme-challenge.#{domain}".sub(/\.\*\./, ".")` via @rudiedirkx patch in https://github.com/jbjonesjr/letsencrypt-manual-hook/issues/10#issuecomment-361069448.

wildcard certificate support: https://community.letsencrypt.org/t/acme-v2-and-wildcard-certificate-support-is-live/55579

What's interesting is that the acme challenge domain drops the asterisk completely to just create a top level certificate. I'll have to give that some 👀

jbjonesjr commented on manual_hook.rb:
```diff
@@ -2,56 +2,75 @@
+    if resp.strings[0] == challenge[:txt_challenge]
+      puts "Found #{resp.strings[0]}. match."
+      return true
+    else
```
remove per @rudiedirkx. not sure why?

pc-fan replied:
I think you are right with https://github.com/jbjonesjr/letsencrypt-manual-hook/issues/10#issuecomment-390383142 the else shouldn't be dropped but we have to iterate over resp.strings

pc-fan replied:
Taking a closer look, L11 already iterates over all TXT records for that domain. We only have to remive the return in the else branch.

jbjonesjr replied:
So what I was originally conveying with this else is that I found a DNS txt record for the site (`_acme_challenge.blahblah.com`), but the value of the TXT record was not valid. This is a different error case than if no TXT record was not found at all.

Maybe this is not relevant, but helps if someone mistyped the value, and is instead trying to figure out "why the TXT record never propagated to the LE servers"

I'll update the error messages to be more clear, but leave the `else` in place.

jbjonesjr reviewed (commented):
BLAST FROM THE PAST.

@pc-fan, thanks for putting this work together, I'm reviewing and testing now to see what it looks like.

Also thanks to @rudiedirkx for his input and updates for acme v2 support. I have an outstanding question on L15, but the rest of this PR looks good.

Sorry it took so long to review, but it `worked on my machine`, and didn't have the bandwidth to reload this context.

I'll go ahead and make the documented changes in the review notes unless someone beats me to it.

pc-fan commented:
Thank you for the review @jbjonesjr 

I'll add my thoughts to the review notes.

pc-fan commented on manual_hook.rb:
```diff
@@ -2,56 +2,75 @@
+      return false
+    end
+  }
+  puts "Didn't find a match for #{challenge[txt_challenge]}"
```
Yes, the colon was missing.

Your error message sounds much clearer, will change it and add the colon.

pc-fan commented on manual_hook.rb:
```diff
@@ -2,56 +2,75 @@
+      puts "Press enter when DNS has been updated..."
+      $stdin.readline
+      first_iteration = false
+    else
```
I always waited till i saw the DNS records propagated (i have set a TTL of 300s for them) and then pressed enter. In that case the 2 minutes were unnecessary for me.

I can revert the change if you like.

jbjonesjr replied:
ohhh, interesting. i let the computer™ confirm it for me, because looking at DNS records is like watching :grass: grow. 😉 At the end of the day, the same thing happens, and you have a valid use case.

👍

pc-fan committed:
Add changes from @jbonesjr and @rudiedirkx for #15 "Enable bulk registration (for use with HOOK_CHAIN=yes)"

jbjonesjr committed:
update error messages

jbjonesjr reviewed (approved):
I need to actual test this from my system on some domains, but I'm 👍 on the code changes at this point

jbjonesjr committed:
fixing errors is hard

diff tells the story of :1000: pains

jbjonesjr committed:
use better words and more consistent error messages

jbjonesjr committed:
clarify error

jbjonesjr commented on manual_hook.rb:
```diff
@@ -2,56 +2,74 @@
+# Check if a challenge is resolved, returns true in that case
+def resolved?(dns, challenge)
+  dns.each_resource(challenge[:acme_domain], Resolv::DNS::Resource::IN::TXT) { |resp|
+    if resp.strings[0] == challenge[:txt_challenge]
```
Still need to loop over resp.strings when an acme_domain has multiple challenges (aka, fqdn and *.fqdn)

jbjonesjr replied:
Confirming my suspicion in https://github.com/jbjonesjr/letsencrypt-manual-hook/issues/10#issuecomment-390383142

jbjonesjr committed:
support multiple values in the txt record

not sure if this will actually work (does the return true escape the each it's in?)

jbjonesjr committed:
fix log message

jbjonesjr committed:
add checkmark, ballot x to stand out

jbjonesjr commented:
test cases:
- goodvalue.fqdn
- badvalue.fqdn
- 2 values that are good, hook_chain=no
- 2 values that are good, hook_chain=yes
- 2 values that are good, fqdn and *.fqdn (these resolve to the same acme_domain

jbjonesjr committed:
clean up logging

jbjonesjr committed:
minor wordsmithing

jbjonesjr reviewed (approved):
got it all done

pc-fan commented:
Thank you :+1:
